### PR TITLE
Fixes #20376 - Add features required for async-ssh

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

File: lib/foreman_tasks_core/runner/action.rb [hunk 1]
```diff
@@ -52,15 +54,34 @@ def finish_run(update)
...
       end
 
+      def merge_outputs(outputs, incremental)
+        # No need to do anything if there are no inputs or if the "new" outputs are older than the already saved ones
+        return if outputs.empty? || (!output[:result].empty? && outputs.first['timestamp'] < output[:result].last['timestamp'])
+        unless incremental
+          # The new output is complete, we need to remove the already known outputs from it
+          # Also since it's complete output, there's only one
+          new = outputs.first
+          known = output[:result].select { |out| out['timestamp'] < outputs.first['timestamp'] }
```
Comment (iNecas) [inline lib/foreman_tasks_core/runner/action.rb]: Couldn't we just iterate though the known and `gsub!(/\A#{out['output]}/)` iteratively, not needed the timestamp at all? I'm wondering, if we can rely on `timestamp`'s uniqueness and if we there are some other risks around that. If we didn't, I would feel a bit better about that.

File: lib/foreman_tasks_core/runner/action.rb [hunk 2]
```diff
@@ -52,15 +54,34 @@ def finish_run(update)
...
         if update.exit_status
           finish_run(update)
         else
           suspend
         end
       end
 
+      def merge_outputs(outputs, incremental)
+        # No need to do anything if there are no inputs or if the "new" outputs are older than the already saved ones
+        return if outputs.empty? || (!output[:result].empty? && outputs.first['timestamp'] < output[:result].last['timestamp'])
```
Comment (iNecas) [inline lib/foreman_tasks_core/runner/action.rb]: the `outputs` is quite confusing with the `output`, something like `new_chunks` and extracting `output[:result]` to `known_chunks`, (or something better than `chunks`:) would help reading this code

File: lib/foreman_tasks_core/runner/base.rb [hunk 3]
```diff
@@ -50,6 +50,19 @@ def timeout
       def timeout_interval
         # A number of seconds after which the runner should receive a #timeout
         #   or nil for no timeout
+        1
+      end
+
+      def external_event(_event)
+        new_data = @continuous_output
+        @continuous_output = ForemanTasksCore::ContinuousOutput.new
+        if !new_data.empty? || @exit_status
```
Comment (iNecas) [inline lib/foreman_tasks_core/runner/base.rb]: This code is very similar to that in `run_refresh`, probably worth extracting

File: lib/foreman_tasks_core/runner/action.rb [hunk 4]
```diff
@@ -52,15 +54,34 @@ def finish_run(update)
...
+        unless incremental
+          # The new output is complete, we need to remove the already known outputs from it
+          # Also since it's complete output, there's only one
+          new = outputs.first
+          known = output[:result].select { |out| out['timestamp'] < outputs.first['timestamp'] }
+                                 .map { |out| out['output'] }.join
+          new['output'].gsub!(/\A#{known}/, '')
```
Comment (iNecas) [inline lib/foreman_tasks_core/runner/action.rb]: The `know` whould be escaped before putting into regexp

File: lib/foreman_tasks_core/runner/base.rb (hunk 3 above)
Comment (iNecas) [inline lib/foreman_tasks_core/runner/base.rb]: ping?

Comment (adamruzicka) [inline lib/foreman_tasks_core/runner/base.rb]: I'm afraid I've lost a commit somewhere...

File: lib/foreman_tasks_core/ticker.rb [hunk 5]
```diff
@@ -1,12 +1,15 @@
 module ForemanTasksCore
   class Ticker < ::Dynflow::Actor
+
+    REFRESH_INTERVAL = 1
```
Comment (iNecas) [inline lib/foreman_tasks_core/ticker.rb]: Is this being used anywhere?

**Comment by iNecas** [comment]
One last comment, tested, and works well, ready to go (after the jenkins passes)

**Comment by adamruzicka** [comment]
Tests are running here http://ci.theforeman.org/job/test_plugin_pull_request/3944/

**Comment by iNecas** [comment]
The re-run tests are running http://ci.theforeman.org/job/test_plugin_pull_request/3945/, looks good so far

**Comment by adamruzicka** [comment]
The tests are almost green, except for [one test](http://ci.theforeman.org/job/test_plugin_pull_request/3945/database=postgresql,label=fast,ruby=2.2/testReport/junit/(root)/PtableIntegrationTest/test_0002_edit_page/)

**Comment by iNecas** [comment]
One last thing: would you mind adding some unit test for the OtpManager? Thanks

**Comment by adamruzicka** [comment]
Added tests, they're running here now http://ci.theforeman.org/job/test_plugin_pull_request/3948/

**Comment by adamruzicka** [comment]
The tests are green now

**Comment by iNecas** [comment]
Merging now, thanks @adamruzicka

**Pull request merged**